### ISSUE TITLE
MIPS PLT entries are broken.

### ISSUE THREAD

**Issue by plafosse** [Vector35/binaryninja-api#3267]
**Version and Platform (required):**
- Binary Ninja Version: 3.1

**Bug Description:**
![image](https://user-images.githubusercontent.com/10556585/178293363-8ac47640-057b-46d6-8bad-a453504a4764.png)

PLT entries in the included binary are broken in some fundamental way.

[libhdk.bndb.zip](https://github.com/Vector35/binaryninja-api/files/9085139/libhdk.bndb.zip)

**Comment by lwerdna** [comment]
This one is kind of tough. First a quick review on what happens when this binary actually executes, here for an example call to memset().

Register $gp gets address `GOT + 0x7FF0`. When the actual call does `lw $t9, -0x7d28($gp)` it resolves to 0x2C6C8, the address of the GOT entry. The entry initially holds the address of the stub. Every stub does `lw $t9, -0x7ff0($gp)`, loading the 0'th entry and calling it. While the binary has bytes 00 00 00 00 there, the dynamic loader at runtime replaces it with a pointer to its lazy loading routine:

![image](https://github.com/Vector35/binaryninja-api/assets/474191/cc87b1fe-d3aa-4869-ab0c-e37505d91d24)

The routine reads $t8 to see which symbol table index it's resolving (notice the `addiu $t8, ...` in the stub's branch slot) and then overwrites the GOT entry with the address of the called function's implementation, which is probably in some shared object that RTLD also was responsible for placing:

![image](https://github.com/Vector35/binaryninja-api/assets/474191/357b7427-7666-43e6-aaad-6a8a51e1a770)

The stub **never executes again**. The initial call has four steps, but every subsequent call only has two steps. What does BinaryNinja do? Well it imitates the loader a bit, by replacing the initial GOT entry (pointing to the stub) with the address of a "resolved" function. Of course BinaryNinja doesn't actually load shared objects dependencies and resolve them, so a fake section ".extern" acts as a stand-in:

![image](https://github.com/Vector35/binaryninja-api/assets/474191/3ee28266-bea1-46cc-940c-4203e729412d)

When the first instruction of each stub loads the first entry of the GOT, it really does get 0x0 because this is a mostly static look at a binary and RTLD hasn't executed and replaced the zero with its lazy loading routine:

```
00018b30  1080998f   lw      $t9, -0x7ff0($gp)  {0x0}  {_GLOBAL_OFFSET_TABLE_}
```

It does end up calling 0x0 through $t9, and thus decompiling to `nullptr()` isn't incorrect.
 
A person wishing to see all callers to a function in this situation should currently navigate to the .got version of it and look at code references, or the .extern version of it which will have one data references from the .got version:

![image](https://github.com/Vector35/binaryninja-api/assets/474191/92d835b6-e720-4aab-b850-2ce3ea03c055)

In other words, ignore the single-use stub for some function `foo()`. Instead look at `foo()`'s GOT entry.

**Comment by lwerdna** [comment]
This finally closes due to:
* 6369988e in binja (improved handling of external pointers)
* 19b21823 in binja (a minimal mipsel32 typelib)
* 0d875b78 in view-elf (create fake RTL_Resolve(), place in GOT with reloc, etc.)
* bf49828d in arch-mips (calling convention for RTL_Resolve())
* 7f38beb4 in arch-mips (function recognizer to relay type of extern to type of plt stub)

Current appearance:
![image](https://github.com/Vector35/binaryninja-api/assets/474191/d27726a2-7622-47fb-870f-db39a8ab1cf0)
